worldengine/draw.py

```diff
@@ -161,14 +163,47 @@ def __setitem__(self, item, value):
         self.pixels[item] = value
 
 
-def draw_simple_elevation(data, width, height, sea_level, target):
+def draw_simple_elevation(world, sea_level, target):
     """ This function can be used on a generic canvas (either an image to save
         on disk or a canvas part of a GUI)
     """
-    for y in range(height):
-        for x in range(width):
-            e = data[y][x]#TODO: numpy-notation [y,x] is a little faster; all passed arrays would have to be numpy-arrays
-            r, g, b = elevation_color(e, sea_level)
+    min_elev_sea = None
+    max_elev_sea = None
+    min_elev_land = None
+    max_elev_land = None
+    for y in range(world.height):
+        for x in range(world.width):
+            e = world.elevation['data'][y,x]
+            if sea_level is None:
+                if min_elev_land is None or e < min_elev_land:
+                    min_elev_land = e
+                if max_elev_land is None or e > max_elev_land:
+                    max_elev_land = e
+            elif world.is_land((x, y)):
+                if min_elev_land is None or e < min_elev_land:
+                    min_elev_land = e
+                if max_elev_land is None or e > max_elev_land:
+                    max_elev_land = e
+            else:
+                if min_elev_sea is None or e < min_elev_sea:
+                    min_elev_sea = e
+                if max_elev_sea is None or e > max_elev_sea:
+                    max_elev_sea = e
+
+    elev_delta_land = (max_elev_land - min_elev_land)/11
+    if sea_level != None:
+        elev_delta_sea = max_elev_sea - min_elev_sea
+    
+    for y in range(world.height):
```

Comment:
The big thing you can do with numpy is operate on the whole array at once.
c is just an array filled with garbage data and the same *shape* as the elevation map (shape is a tuple ```(height, width)``` in this case)
If a numpy array is passed an "index" that is actually a bool array, you will operate on all elements of the array where the booleans are ```True```. So ```c[numpy.invert(world.ocean)]``` operates on all areas of the array c that correspond to land (```c[world.ocean]``` then operates on all that corresponds to ocean).

The things on the right side of the ```=``` are operations that are done on the whole array, as said initially.

The block of code I posted replaces this part:
```diff
+    for y in range(world.height):
+        for x in range(world.width):
+            e = world.elevation['data'][y, x]
+            if sea_level is None:
+                c = ((e - min_elev_land) / elev_delta_land) + 1
+            elif world.is_land((x, y)):
+                c = ((e - min_elev_land) / elev_delta_land) + 1
+            else:
+                c = ((e - min_elev_sea) / elev_delta_sea)
```
Although afterwards you will still have to iterate over the colour-array c to write out its values:
```diff
+    for y in range(world.height):
+        for x in range(world.width):
+            r, g, b = elevation_color(c, sea_level)
             target.set_pixel(x, y, (int(r * 255), int(g * 255),
                                     int(b * 255), 255))
```